eclair-core/src/main/scala/fr/acinq/eclair/Features.scala

```diff
@@ -56,6 +56,12 @@ trait InitFeature extends Feature
 trait NodeFeature extends Feature
 /** Feature that should be advertised in invoices. */
 trait InvoiceFeature extends Feature
+/**
+ * Feature negotiated when opening a channel that are *not* part of the channel type, but will apply for all of the
+ * channel's lifetime. This doesn't include features that can be safely activated/deactivated without impacting the
+ * channel's operation such as option_dataloss_protect or option_shutdown_anysegwit.
+ */
+trait PermanentOptionalChannelFeature extends InitFeature // <- not in the spec
```

Comment:
Because these features must have been set in `init` to be negotiated for a channel opening flow, does that make sense? For example there would be really no reason to have an invoice-only feature in there such as `option_payment_metadata`.

Reply:
I tried an alternative approach in 8fd90554d, where I positively define features:

```scala
trait PermanentChannelFeature extends InitFeature
trait ChannelTypeFeature extends PermanentChannelFeature
```

This adds more typing and forces us to identify channel-type features at definition:

```scala
// permanent feature, but not channel-type
case object DataLossProtect extends Feature with InitFeature with NodeFeature with PermanentChannelFeature
// channel-type feature
case object StaticRemoteKey extends Feature with InitFeature with NodeFeature with ChannelTypeFeature
```

```scala
sealed trait ChannelType {
    def features: Set[_ <: InitFeature]
}
sealed trait SupportedChannelType extends ChannelType { 
    override def features: Set[ChannelTypeFeature] // <- has to be ChannelTypeFeature
}
case class UnsupportedChannelType(featureBits: Features[InitFeature]) extends ChannelType {
    override def features: Set[InitFeature] = featureBits.activated.keySet // <- any type of init feature
}
```

Reply:
That's looking good, thanks for trying it out, I'll apply that.

Reply:
Done and rebased in https://github.com/ACINQ/eclair/pull/2282/commits/64dec7dc1be41a62247e51583664a64daff486a2

Reply:
> That's looking good, thanks for trying it out, I'll apply that.

I see you didn't cherry pick the change to the `ChannelFeatures`signature, is there a reason? I think it makes a lot of sense to have `PermanentChannelFeature` make the most sense there:

```scala
/**
 * Subset of Bolt 9 features used to configure a channel and applicable over the lifetime of that channel.
 * Even if one of these features is later disabled at the connection level, it will still apply to the channel until the
 * channel is upgraded or closed.
 */
ChannelFeatures(features: Set[PermanentChannelFeature])
```

> Note that I didn't include DataLossProtect in the permanent channel features, but I can include it if you think it makes more sense.

I was not sure about `DataLossProtect`. In theory it should be permanent but that's not enforceable.

Reply:
Let's start without it which is what we were previously doing, we can add it later if we think it's important!